### PR TITLE
Update ctr_model.py

### PR DESCRIPTION
**Pull request by DSTeam109**

### CODE REVIEW

File: ctr_model.py [hunk 1]
```diff
@@ -1 +1,157 @@
...
+plt.title('Normalized missing data count per column')
+plt.xlim(0, 1)
+
+numeric_columns = ['position', 'gig_price', 'gig_avg_rating', 'gig_rated_orders', 'cnt_helpful_reviews_last_year', 'avg_review_length_last_3_months',
+       'avg_position_shown_last_60d', 'previous_order_amount']
+dates_columns = ['created_at', 'user_reg_date', 'previous_order_date']
+binary_columns = ['is_click', 'is_filtered', 'is_user_buyer','is_seller_onlie']
+categorical_columns = list(set(full_df.columns) - set(numeric_columns) - set(dates_columns) - set(binary_columns))
+
+full_df['continent'] = full_df['user_timezone'].apply(lambda x: x.split('/')[0].lower())
```
Comment (EyalBustani) [inline ctr_model.py:27]: full_df['continent'] = full_df['user_timezone']**.astype(str)**.apply(lambda x: x.split('/')[0].lower())

File: ctr_model.py [hunk 2]
```diff
@@ -1 +1,157 @@
...
+    df['user_reg_date'] = pd.to_datetime(df['user_reg_date'])
+    df['previous_order_date'] = pd.to_datetime(df['previous_order_date'])
+
+    df['hour'] = (df.created_at.dt.hour + (df.created_at.dt.minute / 60)).round(1)
+    df['day'] = df.created_at.dt.day_name().apply(lambda x: x.lower()[:3])
+    df['week_in_month'] = df.created_at.dt.day // 7
+    df['months_since_reg'] = ((df.created_at - df.user_reg_date).dt.days / 30.2).round(1)
+    df['hour_diff_from_last_purchase'] = (df.created_at.dt.hour - df.previous_order_date.dt.hour).round(1)
+    df['days_since_last_purchase'] = (df.created_at - df.previous_order_date).dt.days
+    #df['is_weekend'] = df.apply(is_weekend, axis=1)
```
Comment (EyalBustani) [inline ctr_model.py:62]: df['is_weekend'] = np.where(df['day'].isin(['fri','sat']), 1,0)

File: ctr_model.py [hunk 3]
```diff
@@ -1 +1,157 @@
...
+    df['price_change_from_last_purchase'] = df.apply(compare_order_amounts, axis=1)
+    df['is_same_sc_as_last_purchase'] = df.apply(compare_sc_id, axis=1)
+    df['position_diff_from_avg'] = df.position - df.avg_position_shown_last_60d
+    return df
+
+full_df = add_gig_features(full_df)
+
+x_train, x_test, y_train, y_test = train_test_split(full_df.drop('is_click', axis=1), full_df['is_click'], test_size=0.2)
+x_train, x_val, y_train, y_val = train_test_split(x_train, y_train, test_size=0.2)
+
```
Comment (EyalBustani) [inline ctr_model.py:98]: I would use a different split for the test set. 
For example, the test should be out-of-time, so you could split the data with "created_at" and take all the most recent rows (i.e. of last week) for your test and the rest would be split for train and validation.

File: ctr_model.py [hunk 4]
```diff
@@ -1 +1,157 @@
...
+    scale_pos_weight=2,
+)
+
+feature_importance = []
+cv_metrics = []
+n_folds = 6
+full_df = full_df.reset_index(drop=True)
+
+splitter = KFold(n_splits=n_folds, shuffle=True, random_state=13)
+for i, (train_indices, test_indices) in enumerate(splitter.split(full_df)):
```
Comment (EyalBustani) [inline ctr_model.py:120]: I'm not sure what the purpose of this Kflod is, it's usually used to validate model performance in case you don't have enough data. If that's the case then why did you split the data into train-valid-test before?

File: ctr_model.py [hunk 5]
```diff
@@ -1 +1,157 @@
...
+    learning_rate=0.2,
+    n_estimators=200, 
+    max_depth=7,
+    colsample_bytree=0.9,
+    colsample_bylevel=0.9,
+    colsample_bynode=0.9,
+    importance_type='cover',
+    scale_pos_weight=2,
+)
+
```
Comment (EyalBustani) [inline ctr_model.py:113]: What about hyperparam optimization?

File: ctr_model.py [hunk 6]
```diff
@@ -1 +1,157 @@
...
+              eval_set=[(x_train, y_train), (x_val, y_val)], verbose=50)
+    y_prob = model.predict_proba(x_test)[:, 1]
+    y_pred = (y_prob > 0.5).astype(int)
+    feature_importance.append(dict(zip(model.feature_names_in_, model.feature_importances_)))
+    cv_metrics.append({
+        'accuracy': metrics.accuracy_score(y_test, y_pred),
+        'balanced_accuracy': metrics.balanced_accuracy_score(y_test, y_pred),
+        'f1': metrics.f1_score(y_test, y_pred),
+        'auc_roc': metrics.roc_auc_score(y_test, y_prob),
+    })
```
Comment (EyalBustani) [inline ctr_model.py:140]: It seems that accuracy and f1 are not the right metrics for an imbalanced classifier (~9%  clicks).
AUC is a good metric for ranking evaluation, and I would also add the log_loss

File: ctr_model.py [hunk 7]
```diff
@@ -1 +1,157 @@
...
+    x_test = full_df.loc[test_indices, :].drop('is_click', axis=1)
+    y_test = full_df['is_click'][test_indices].astype(int)
+    
+    x_train, x_val, y_train, y_val = train_test_split(x_train, y_train, test_size=0.2, stratify=y_train, random_state=13)
+    
+    model = xgb.XGBClassifier(eval_metric='logloss', enable_categorical=True, **model_params)
+    model.fit(x_train, y_train,
+              eval_set=[(x_train, y_train), (x_val, y_val)], verbose=50)
+    y_prob = model.predict_proba(x_test)[:, 1]
+    y_pred = (y_prob > 0.5).astype(int)
```
Comment (EyalBustani) [inline ctr_model.py:133]: The data is imbalanced (9% is_click), why  y_prob > 0.5 ?

File: ctr_model.py [hunk 8]
```diff
@@ -1 +1,157 @@
...
+numeric_columns = ['position', 'gig_price', 'gig_avg_rating', 'gig_rated_orders', 'cnt_helpful_reviews_last_year', 'avg_review_length_last_3_months',
+       'avg_position_shown_last_60d', 'previous_order_amount']
+dates_columns = ['created_at', 'user_reg_date', 'previous_order_date']
+binary_columns = ['is_click', 'is_filtered', 'is_user_buyer','is_seller_onlie']
+categorical_columns = list(set(full_df.columns) - set(numeric_columns) - set(dates_columns) - set(binary_columns))
+
+full_df['continent'] = full_df['user_timezone'].apply(lambda x: x.split('/')[0].lower())
+vc = full_df.continent.value_counts()
+vc.plot(kind='bar')
+
```
Comment (EyalBustani) [inline ctr_model.py:30]: About 96% of the data are from "America" continent, seems like you are losing information by using this feature instead of user_timezone. I would consider at least adding user_timezone as a feature.

File: ctr_model.py [hunk 9]
```diff
@@ -1 +1,157 @@
...
+    values_to_replace = value_counts[value_counts < thresh].index
+    return ser.replace(values_to_replace, 'other')
+
+for col in categorical_columns:
+    full_df[col] = rare_cats_to_other(full_df[col])
+    full_df[col].fillna('other', inplace=True)
+
+for col in numeric_columns:
+    full_df[col].fillna(0, inplace=True)
+
```
Comment (EyalBustani) [inline ctr_model.py:79]: Not sure if 0 as a default for every numeric feature is the correct imputation. 
For example, promoted impressions don't have gig_avg_rating, and imputed 0 could cause a bias in the prediction.
